internal/controllers/machineset/machineset_controller_status.go

```diff
@@ -0,0 +1,314 @@
+
+	// MachinesReady condition: aggregate the Machine's Ready condition
+	setMachinesReadyCondition(s.machineSet, s.machines)
+
+	// MachinesUpToDate condition: aggregate the Machine's UpToDate condition
+	setMachinesUpToDateCondition(s.machineSet, s.machines)
+
+	// TODO Deleting
+}
+
+func setReplicas(ms *clusterv1.MachineSet, machines []*clusterv1.Machine) {
+	// Return early when machines is nil. It's not possible to calculate replica counters.
+	// Conditions may surface to be unknown in this case.
+	if machines == nil {
+		return
+	}
+
+	var readyReplicas, availableReplicas int32
+	for _, machine := range machines {
+		if v1beta2conditions.IsTrue(machine, clusterv1.MachineReadyV1Beta2Condition) {
+			readyReplicas++
+		}
+		if v1beta2conditions.IsTrue(machine, clusterv1.MachineAvailableV1Beta2Condition) {
+			availableReplicas++
+		}
+	}
+
+	if ms.Status.V1Beta2 == nil {
+		ms.Status.V1Beta2 = &clusterv1.MachineSetV1Beta2Status{}
+	}
+
+	ms.Status.V1Beta2.ReadyReplicas = ptr.To(readyReplicas)
+	ms.Status.V1Beta2.AvailableReplicas = ptr.To(availableReplicas)
+}
+
+func setScalingUpCondition(ms *clusterv1.MachineSet, machines []*clusterv1.Machine, bootstrapObjectNotFound, infrastructureObjectNotFound bool) {
+	// If we got unexpected errors in listing the machines (this should happen rarely), surface them
+	if machines == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineSetScalingUpInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	// Surface if .spec.replicas is not yet set (this should happen rarely).
+	// This could e.g. be the case when using autoscaling with MachineSets.
+	if ms.Spec.Replicas == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:   clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineSetScalingUpWaitingForReplicasToBeSetV1Beta2Reason,
+		})
+		return
+	}
+
+	desiredReplicas := *ms.Spec.Replicas
+	if !ms.DeletionTimestamp.IsZero() {
+		desiredReplicas = 0
+	}
+	gotReplicas := ms.Status.Replicas
+
+	referencesMessage := calculateMissingReferencesMessage(ms, bootstrapObjectNotFound, infrastructureObjectNotFound)
+
+	// Not scaling up or in deletion.
+	if gotReplicas >= desiredReplicas {
+		var message string
+		if referencesMessage != "" {
+			message = fmt.Sprintf("ScalingUp can't happen %s", referencesMessage)
+		}
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionFalse,
+			Reason:  clusterv1.MachineSetNotScalingUpV1Beta2Reason,
+			Message: message,
+		})
+		return
+	}
+
+	message := fmt.Sprintf("ScalingUp from %d to %d replicas", gotReplicas, desiredReplicas)
```

Comment:
```suggestion
	message := fmt.Sprintf("Scaling up from %d to %d replicas", gotReplicas, desiredReplicas)
```
Let's use regular english in the message, this seems a bit awkward otherwise (please also check the other messages)